File: cli/internal/checks/dnsbl_codes.go
package checks

import (
	"os"
	"strings"
)

type DnsblKind string

const (
	DnsblClean      DnsblKind = "clean"
	DnsblListed     DnsblKind = "listed"
	DnsblWhitelist  DnsblKind = "whitelist"
	DnsblQueryError DnsblKind = "query_error"
)

type DnsblInterp struct {
	Kind   DnsblKind
	Status Severity
	Label  string
	Detail string
}

var spamhausListed = map[string]string{
	"127.0.0.2":  "SBL — spam source / snowshoe",
	"127.0.0.3":  "CSS — snowshoe / exploited",
	"127.0.0.4":  "XBL/CBL — exploited/compromised host",
	"127.0.0.5":  "XBL — reserved listing code",
	"127.0.0.6":  "XBL — reserved listing code",
	"127.0.0.7":  "XBL — reserved listing code",
	"127.0.0.9":  "SBL DROP / hijacked",
	"127.0.0.10": "PBL — ISP dynamic/end-user",
	"127.0.0.11": "PBL — ISP dynamic/end-user",
	"127.0.0.30": "BCL — botnet controller",
}

var cblListed = map[string]string{
	"127.0.0.2": "CBL — exploited host",
	"127.0.0.4": "CBL/XBL — exploited host",
}

var sorbsListed = map[string]string{
	"127.0.0.2":  "HTTP",
	"127.0.0.3":  "SOCKS",
	"127.0.0.4":  "MISC",
	"127.0.0.5":  "SMTP",
	"127.0.0.6":  "WEB",
	"127.0.0.7":  "BLOCK",
	"127.0.0.8":  "ZOMBIE",
	"127.0.0.9":  "DUL (dynamic)",
	"127.0.0.10": "BADCONF",
	"127.0.0.11": "NOSERVER",
}

var queryErrors = map[string]string{
	"127.255.255.252": "Malformed DNSBL zone name.",
	"127.255.255.254": "Blocked as an anonymous/open-resolver query. Public resolvers cannot use Spamhaus public mirrors. Set SPAMHAUS_DQS_KEY for a private DQS query. This is not a listing.",
	"127.255.255.255": "Excessive queries — rate limited. Not a listing.",
}

const listingNotError = "This is a listing return code (127.0.0.0/24), not a query error. Query errors are only 127.255.255.0/24."

func listingMap(zone string) map[string]string {
	z := strings.ToLower(zone)
	switch {
	case strings.Contains(z, "spamhaus"):
		return spamhausListed
	case strings.Contains(z, "abuseat"), strings.Contains(z, "cbl."):
		return cblListed
	case strings.Contains(z, "sorbs"):
		return sorbsListed
	default:
		return nil
	}
}

func InterpretDnsbl(zone string, answers []string, whitelist bool) DnsblInterp {
	if len(answers) == 0 {
		return DnsblInterp{Kind: DnsblClean, Status: StatusOK, Label: "OK", Detail: "Not listed"}
	}
	for _, a := range answers {
		if detail, ok := queryErrors[a]; ok || strings.HasPrefix(a, "127.255.255.") {
			if !ok {
				detail = "DNSBL returned error code " + a + " (127.255.255.0/24). This is not a reputation listing."
			}
			if strings.Contains(zone, "spamhaus") {
				detail += " For private deploys, set SPAMHAUS_DQS_KEY so queries use {key}.zen.dq.spamhaus.net."
			}
			return DnsblInterp{Kind: DnsblQueryError, Status: StatusWarn, Label: "Query error " + a, Detail: detail}
		}
	}
	if whitelist {
		return DnsblInterp{Kind: DnsblWhitelist, Status: StatusOK, Label: "Listed (good) " + strings.Join(answers, ", "), Detail: "Present on a DNS whitelist."}
	}
	mp := listingMap(zone)
	labels := make([]string, 0, len(answers))
	for _, a := range answers {
		if mp != nil {
			if name, ok := mp[a]; ok {
				labels = append(labels, a+" ("+name+")")
				continue
			}
		}
		labels = append(labels, a)
	}
	detail := "Present on this DNSBL. " + listingNotError
	if mp != nil {
		if name, ok := mp[answers[0]]; ok {
			detail = name + ". " + listingNotError
		}
	}
	return DnsblInterp{Kind: DnsblListed, Status: StatusFail, Label: "LISTED " + strings.Join(labels, ", "), Detail: detail}
}

func SpamhausZone() (zone, name string) {
	key := strings.TrimSpace(os.Getenv("SPAMHAUS_DQS_KEY"))
	if key != "" {
		return key + ".zen.dq.spamhaus.net", "Spamhaus ZEN (DQS)"
	}
	return "zen.spamhaus.org", "Spamhaus ZEN"
}

File: worker/checks/dnsbl-codes.ts
import type { Severity } from './types';

/** DNSBL answers in 127.255.255.0/24 are query errors, not listings. */
export type DnsblKind = 'clean' | 'listed' | 'whitelist' | 'query_error';

export type DnsblInterpretation = {
	kind: DnsblKind;
	status: Severity;
	label: string;
	detail: string;
};

const SPAMHAUS_LISTED: Record<string, string> = {
	'127.0.0.2': 'SBL — spam source / snowshoe',
	'127.0.0.3': 'CSS — snowshoe / exploited',
	'127.0.0.4': 'XBL/CBL — exploited host',
	'127.0.0.9': 'SBL DROP / hijacked',
	'127.0.0.10': 'PBL — ISP dynamic/end-user',
	'127.0.0.11': 'PBL — ISP dynamic/end-user',
};

const QUERY_ERRORS: Record<string, string> = {
	'127.255.255.252': 'Malformed DNSBL zone name.',
	'127.255.255.254':
		'Blocked as an anonymous/open-resolver query. Public resolvers (including Cloudflare 1.1.1.1) cannot use Spamhaus public mirrors. Add a Spamhaus DQS key on a private instance, or this is not a listing.',
	'127.255.255.255': 'Excessive queries — rate limited. Not a listing.',
};

export function isQueryErrorCode(ip: string): boolean {
	return ip.startsWith('127.255.255.');
}

export function interpretDnsblCodes(zone: string, answers: string[], whitelist = false): DnsblInterpretation {
	if (!answers.length) {
		return { kind: 'clean', status: 'ok', label: 'OK', detail: 'Not listed' };
	}

	const error = answers.find((a) => QUERY_ERRORS[a] || isQueryErrorCode(a));
	if (error) {
		const detail =
			QUERY_ERRORS[error] ??
			`DNSBL returned error code ${error} (127.255.255.0/24). This is not a reputation listing.`;
		return {
			kind: 'query_error',
			status: 'warn',
			label: `Query error ${error}`,
			detail: zone.includes('spamhaus')
				? `${detail} For private deploys, set SPAMHAUS_DQS_KEY (or paste it in Settings) and queries go to {key}.zen.dq.spamhaus.net.`
				: detail,
		};
	}

	if (whitelist) {
		return {
			kind: 'whitelist',
			status: 'ok',
			label: `Listed (good) ${answers.join(', ')}`,
			detail: 'Present on a DNS whitelist.',
		};
	}

	const labels = answers.map((a) => {
		if (zone.includes('spamhaus') && SPAMHAUS_LISTED[a]) return `${a} (${SPAMHAUS_LISTED[a]})`;
		return a;
	});

	return {
		kind: 'listed',
		status: 'fail',
		label: `LISTED ${labels.join(', ')}`,
		detail: zone.includes('spamhaus') ? 'Spamhaus dataset hit — see return code meaning above.' : 'Present on this DNSBL.',
	};
}

/** Public zen.spamhaus.org rejects DoH/public resolvers. DQS uses a keyed zone. */
export function spamhausZone(dqsKey?: string): { zone: string; name: string; url: string } {
	const key = dqsKey?.trim();
	if (key) {
		return {
			zone: `${key}.zen.dq.spamhaus.net`,
			name: 'Spamhaus ZEN (DQS)',
			url: 'https://www.spamhaus.org/lookup/',
		};
	}
	return {
		zone: 'zen.spamhaus.org',
		name: 'Spamhaus ZEN',
		url: 'https://www.spamhaus.org/lookup/',
	};
}
